src/models/Offer.js

```diff
@@ -1,21 +1,50 @@
-export default class Offer {
-  constructor ({id, email, city, formData, expirationDate, confirmed, createdDate, deleted, hashedToken}) {
-    this.id = id
-    this.email = email
-    this.city = city
-    this.formData = formData
-    this.expirationDate = expirationDate
-    this.confirmed = confirmed
-    this.createdDate = createdDate
-    this.deleted = deleted
-    this.hashedToken = hashedToken
-  }
+import mongoose from 'mongoose'
+import forms from './forms'
 
-  isExpired () {
-    return this.expirationDate <= Date.now()
+const offerSchema = mongoose.Schema({
+  email: {
+    type: String,
+    lowercase: true,
+    required: [true, 'Missing email']
+  },
+  city: {
+    type: String,
+    lowercase: true,
+    required: [true, 'Missing city']
+  },
+  expirationDate: {
+    type: Date,
+    required: [true, 'Missing expiration date']
+  },
+  confirmed: {
+    type: Boolean,
+    required: [true, 'Missing confirmed property'],
+    default: false
+  },
+  deleted: {
+    type: Boolean,
+    required: [true, 'Missing deleted property'],
+    default: false
+  },
+  createdDate: {
+    type: Date,
+    required: [true, 'Missing created date'],
+    default: Date.now
+  },
+  hashedToken: {
+    type: String,
+    required: [true, 'Missing hashed token'],
+    unique: true // Create a mongodb unique index
+  },
+  formData: {
+    type: mongoose.Schema.Types.ObjectId,
+    ref: Array.from(Object.keys(forms)), // Must be one of the forms defined in forms.js
+    required: [true, 'Missing form data']
```

Comment:
Do we have to write these error messages (for required fields) ourselves? Is mongoose able to create these generic messages itself? 🤔

Reply:
As far as I know mongoose is able to tell at which item something is wrong, so i.e. that something is wrong with the email. Writing custom error messages is optional .